File: examples/java/src/main/java/org/example/objectstore/Operations.java
package org.example.objectstore;

import com.amazonaws.ClientConfiguration;
import com.amazonaws.SDKGlobalConfiguration;
import com.amazonaws.auth.AWSCredentials;
import com.amazonaws.auth.AWSStaticCredentialsProvider;
import com.amazonaws.client.builder.AwsClientBuilder;
import com.amazonaws.regions.Regions;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.Bucket;
import com.amazonaws.services.s3.model.ObjectListing;
import com.amazonaws.services.s3.model.ObjectMetadata;
import com.amazonaws.services.s3.model.PutObjectRequest;
import com.amazonaws.services.s3.model.S3Object;
import com.amazonaws.services.s3.model.S3ObjectSummary;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;

public class Operations {
  private static final String S3_SIGNER = "S3SignerType";
  private static final String S3_V4_SIGNER = "AWSS3V4SignerType";

  public static AmazonS3 getConnection(String url, AWSCredentials credentials, boolean isS3v4) {
      final ClientConfiguration clientConfiguration = new ClientConfiguration();
      clientConfiguration.setSignerOverride(isS3v4 ? S3_V4_SIGNER : S3_SIGNER);

      System.setProperty(SDKGlobalConfiguration.DISABLE_CERT_CHECKING_SYSTEM_PROPERTY, "true");

      return AmazonS3ClientBuilder
        .standard()
        .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(url, Regions.US_EAST_1.name()))
        .withPathStyleAccessEnabled(true)
        .withClientConfiguration(clientConfiguration)
        .withCredentials(new AWSStaticCredentialsProvider(credentials))
        .build();
  }

  public static void demo(AmazonS3 s3, String bucketName, String file) {
    System.out.println();
    Operations.listBuckets(s3);

    System.out.println();
    Operations.createBucketIfNotExists(s3, bucketName);
    Operations.listBuckets(s3);

    System.out.println();
    Operations.uploadFileIfNotExists(s3, bucketName, file);
    Operations.listFolder(s3, bucketName);

    System.out.println();
    Operations.readFile(s3, bucketName, file);

    System.out.println();
    Operations.clearBucket(s3, bucketName);
    Operations.listFolder(s3, bucketName);

    System.out.println();
    Operations.deleteBucketIfExists(s3, bucketName);
    Operations.listBuckets(s3);
  }

  public static void listBuckets(AmazonS3 s3) {
    List<Bucket> buckets = s3.listBuckets();
    System.out.println("Buckets:");
    buckets.stream().map(Bucket::getName).forEach(System.out::println);
  }

  public static boolean checkIfBucketExists(AmazonS3 s3, String bucketName) {
    System.out.println("Checking if bucket '" + bucketName + "' exists...");
    return s3.doesBucketExistV2(bucketName);
  }

  public static void createBucketIfNotExists(AmazonS3 s3, String bucketName) {
    if (checkIfBucketExists(s3, bucketName)) {
      System.out.println("Bucket '" + bucketName + "' exists");
      return;
    }

    System.out.println("Creating bucket '" + bucketName + "' ...");
    s3.createBucket(bucketName);
  }

  public static void deleteBucketIfExists(AmazonS3 s3, String bucketName) {
    if (!checkIfBucketExists(s3, bucketName)) {
      System.out.println("Bucket '" + bucketName + "' not exists");
      return;
    }

    System.out.println("Deleting bucket '" + bucketName + "' ...");
    s3.deleteBucket(bucketName);
  }

  public static boolean checkIfFileExists(AmazonS3 s3, String bucketName, String file) {
    System.out.println("Checking if file '" + file + "' exists in bucket '" + bucketName + "'");
    return s3.doesObjectExist(bucketName, file);
  }

  public static void uploadFileIfNotExists(AmazonS3 s3, String bucketName, String file) {
    if (checkIfFileExists(s3, bucketName, file)) {
      System.out.println("File '" + file + "' exists in bucket '" + bucketName + "'");
      return;
    }

    SimpleDateFormat formatter= new SimpleDateFormat("MM/dd/yyyy 'at' HH:mm:ss");
    Date date = new Date(System.currentTimeMillis());
    String data = "Hello world on " + formatter.format(date) + "!";
    System.out.println("Uploading '" + file + "' to '" + bucketName + "' with data:\n" + data + "\n");

    ObjectMetadata metadata = new ObjectMetadata();
    metadata.setContentType("plain/text");

    InputStream dataStream = new ByteArrayInputStream(data.getBytes());
    PutObjectRequest request = new PutObjectRequest(bucketName, file, dataStream, metadata);

    s3.putObject(request);
  }

  public static void listFolder(AmazonS3 s3, String bucketName) {
    System.out.println("Listing bucket '" + bucketName +"':");
    ObjectListing listing = s3.listObjects(bucketName);
    listing.getObjectSummaries()
      .stream()
      .map(S3ObjectSummary::getKey)
      .forEach(System.out::println);
  }

  public static void clearBucket(AmazonS3 s3, String bucketName) {
    System.out.println("Cleaning all from bucket '" + bucketName +"'..");
    ObjectListing listing = s3.listObjects(bucketName);
    listing.getObjectSummaries()
      .stream()
      .map(S3ObjectSummary::getKey)
      .forEach(file -> s3.deleteObject(bucketName, file));
  }

  public static void readFile(AmazonS3 s3, String bucketName, String file) {
    System.out.println("Reading file '" + file +"' from bucket '" + bucketName + "'");
    S3Object object = s3.getObject(bucketName, file);
    String data = new BufferedReader(new InputStreamReader(object.getObjectContent(), StandardCharsets.UTF_8))
        .lines()
      .collect(Collectors.joining("\n"));
    System.out.println("Data:\n" + data);
  }
}

File: examples/python/operations.py
from datetime import datetime

from botocore.exceptions import ClientError


def demo(s3, bucket_name, file):
    print("\n")
    list_buckets(s3)

    print("\n")
    create_bucket_if_not_exists(s3, bucket_name)
    list_buckets(s3)

    print("\n")
    upload_file_if_not_exists(s3, bucket_name, file)
    list_folder(s3, bucket_name)

    print("\n")
    read_file(s3, bucket_name, file)

    print("\n")
    delete_all(s3, bucket_name)
    list_folder(s3, bucket_name)

    print("\n")
    delete_bucket_if_exists(s3, bucket_name)
    list_buckets(s3)

def list_buckets(s3):
    print("Buckets:")
    for bucket in s3.list_buckets()['Buckets']:
        print(bucket['Name'])


def create_bucket_if_not_exists(s3, bucket_name):
    if check_if_bucket_exists(s3, bucket_name):
        print("Bucket exists")
        return
    print("Bucket does not exist, creating...")
    s3.create_bucket(Bucket=bucket_name)


def check_if_bucket_exists(s3, bucket_name):
    print("Checking if bucket '" + bucket_name + "' exists")
    try:
        s3.head_bucket(Bucket=bucket_name)
    except ClientError as e:
        return int(e.response['Error']['Code']) != 404
    return True


def delete_bucket_if_exists(s3, bucket_name):
    print("Checking if bucket '" + bucket_name + "' exists")
    try:
        s3.head_bucket(Bucket=bucket_name)
        print("Bucket exists, deleting...")
        s3.delete_bucket(Bucket=bucket_name)
    except ClientError:
        print("Bucket does not exist")


def list_folder(s3, bucket_name):
    print("Listing bucket " + bucket_name + ":")
    objects = s3.list_objects(Bucket=bucket_name)
    if 'Contents' in objects:
        for key in objects['Contents']:
            print(key['Key'])
    else:
        print("Is empty")


def delete_all(s3, bucket_name):
    print("Cleaning all form bucket: " + bucket_name)
    objects = s3.list_objects(Bucket=bucket_name)
    if 'Contents' in objects:
        for key in objects['Contents']:
            s3.delete_object(Bucket=bucket_name, Key=key['Key'])


def upload_file_if_not_exists(s3, bucket_name, file_name):
    if check_if_file_exist(s3, bucket_name, file_name):
        print("File '" + file_name + "' already exists in bucket '" + bucket_name + "'")
        return

    date = datetime.now().strftime("%m/%d/%Y, %H:%M:%S")
    data = "Hello world on " + date + "!"

    print("Uploading '" + file_name + "' to '" + bucket_name + "' with data:\n" + data)
    s3.put_object(Body=data.encode('ascii'), Bucket=bucket_name, Key=file_name)


def read_file(s3, bucket_name, file_name):
    print("Reading file '" + file_name + "' from bucket '" + bucket_name + "'")
    obj = s3.get_object(Bucket=bucket_name, Key=file_name)
    print("Data:")
    print(obj['Body'].read().decode('utf-8'))


def check_if_file_exist(s3, bucket_name, file):
    print("Checking if file '" + file + "' exists in bucket '" + bucket_name + "'")
    try:
        s3.head_object(Bucket=bucket_name, Key=file)
    except ClientError as e:
        return int(e.response['Error']['Code']) != 404
    return True
